File: golang/matcher/record.go
package matcher

import (
	"encoding/json"
	"strconv"
	"strings"
)

type Record struct {
	data map[string]any
}

func NewRecord(data map[string]any) *Record {
	return &Record{data: data}
}

func isProbablyJSONString(value any) bool {
	str, ok := value.(string)
	if !ok {
		return false
	}
	if strings.HasPrefix(str, "{") && strings.HasSuffix(str, "}") {
		return true
	}
	if strings.HasPrefix(str, "[") && strings.HasSuffix(str, "]") {
		return true
	}
	return false
}

func extractPath(value any, path []string) any {
	current := value
	for _, key := range path {
		switch c := current.(type) {
		case []any:
			idx, err := strconv.Atoi(key)
			if err != nil || idx < 0 || idx >= len(c) {
				return nil
			}
			current = c[idx]
		case map[string]any:
			v, exists := c[key]
			if !exists {
				return nil
			}
			current = v
		default:
			return nil
		}
	}
	return current
}

func (r *Record) GetValue(key Key) any {
	value, exists := r.data[key.Value]
	if !exists {
		return nil
	}

	if len(key.Path) == 0 {
		return value
	}

	if isProbablyJSONString(value) {
		var parsed any
		if err := json.Unmarshal([]byte(value.(string)), &parsed); err != nil {
			return nil
		}
		return extractPath(parsed, key.Path)
	}

	m, ok := value.(map[string]any)
	if !ok {
		return nil
	}
	return extractPath(m, key.Path)
}

File: javascript/packages/flyql/src/matcher/record.js
import { parseKey } from '../core/key.js'

function isProbablyJSONString(value) {
    if (typeof value !== 'string') return false
    if (value.startsWith('{') && value.endsWith('}')) return true
    if (value.startsWith('[') && value.endsWith(']')) return true
    return false
}

function extractPath(value, path) {
    let current = value
    for (const key of path) {
        if (current === null || current === undefined || typeof current !== 'object') return null
        if (Array.isArray(current)) {
            const idx = parseInt(key, 10)
            if (isNaN(idx) || idx < 0 || idx >= current.length) return null
            current = current[idx]
        } else {
            if (!(key in current)) return null
            current = current[key]
        }
    }
    return current
}

export class Record {
    constructor(data) {
        this.data = data
    }

    getValue(rawKey) {
        const key = parseKey(rawKey)
        const rootKey = key.segments[0]
        const path = key.segments.slice(1)

        const value = this.data[rootKey]
        if (value === undefined) return null

        if (path.length === 0) return value

        if (isProbablyJSONString(value)) {
            try {
                const parsed = JSON.parse(value)
                return extractPath(parsed, path)
            } catch {
                return null
            }
        }

        if (typeof value === 'object' && value !== null) {
            return extractPath(value, path)
        }

        return null
    }
}
